test/jdk/java/util/zip/ZipFile/InvalidBytesInEntryNameOrComment.java

```diff
@@ -0,0 +1,182 @@
+        invalidName = invalidName("invalid-name.zip", templateZip);
+
+        // Create a ZIP with a CEN comment field containing an invalid byte sequence
+        invalidComment = invalidComment("invalid-comment.zip", templateZip);
+    }
+
+    /**
+     * Opening a ZipFile with an invalid UTF-8 byte sequence in
+     * the name field of a CEN file header should throw a
+     * ZipException with "bad entry name or comment"
+     */
+    @Test(expectedExceptions = ZipException.class,
+            expectedExceptionsMessageRegExp = BAD_ENTRY_NAME_OR_COMMENT)
+    public void shouldRejectInvalidName() throws IOException {
+        try (ZipFile zf = new ZipFile(invalidName.toFile())) {
+            // Should throw ZipException
+        }
+    }
+
+    /**
+     * Opening a ZipFile with an invalid UTF-8 byte sequence in
+     * the comment field of a CEN file header should throw a
+     * ZipException with "bad entry name or comment"
+     */
+    @Test(expectedExceptions = ZipException.class,
+            expectedExceptionsMessageRegExp = BAD_ENTRY_NAME_OR_COMMENT)
+    public void shouldIgnoreInvalidComment() throws IOException {
+        try (ZipFile zf = new ZipFile(invalidComment.toFile())) {
+            // Should throw ZipException
+        }
+    }
+
+    /**
+     * Make a valid ZIP file used as a template for invalid files
+     */
+    private byte[] templateZIP() throws IOException {
+        ByteArrayOutputStream bout = new ByteArrayOutputStream();
+        try (ZipOutputStream zo = new ZipOutputStream(bout)) {
+            ZipEntry commentEntry = new ZipEntry("file");
+            commentEntry.setComment("Comment");
+            zo.putNextEntry(commentEntry);
+        }
+        return bout.toByteArray();
+    }
+
+    /**
+     * Make a ZIP with invalid bytes in the CEN name field
+     */
+    private Path invalidName(String name, byte[] template) throws IOException {
+        ByteBuffer buffer = copyTemplate(template);
+        int off = cenStart(buffer);
+        // Name field starts here
+        int noff = off + CEN_HDR;
+
+        // Write invald bytes
+        buffer.put(noff, INVALID_UTF8_BYTE_SEQUENCE, 0, INVALID_UTF8_BYTE_SEQUENCE.length);
+        return writeFile(name, buffer);
+
+    }
+
+    /**
+     * Make a copy of the ZIP template and wrap it in a little-endian
+     * ByteBuffer
+     */
+    private ByteBuffer copyTemplate(byte[] template) {
+        return ByteBuffer.wrap(Arrays.copyOf(template, template.length))
+                .order(ByteOrder.LITTLE_ENDIAN);
+    }
+
+    /**
+     * Make a ZIP with invalid bytes in the CEN comment field
+     */
+    private Path invalidComment(String name, byte[] template) throws IOException {
+        ByteBuffer buffer = copyTemplate(template);
+        int off = cenStart(buffer);
+        // Need to skip past the length of the name and extra fields
+        int nlen = buffer.getShort(off + NLEN);
+        int elen = buffer.getShort(off + NLEN);
```

Comment:
Is this supposed to say `ELEN`?

Reply:
Indeed, good catch! Must have lucked out on the comment be longer than the name :-)